lib/rules/max-classes-per-file.js

```diff
@@ -0,0 +1,80 @@
+/**
+ * @fileoverview Enforce a maximum number of classes per file
+ * @author James Garbutt <https://github.com/43081j>
+ */
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        docs: {
+            description: "enforce a maximum number of classes per file",
+            category: "Best Practices",
+            recommended: false,
+            url: "https://eslint.org/docs/rules/max-classes-per-file"
+        },
+        schema: [
+            {
+                type: "integer",
+                minimum: 1
+            },
+            {
+                type: "object",
+                properties: {
+                    includeClassExpressions: {
+                        type: "boolean"
+                    }
+                },
+                additionalProperties: false
+            }
+        ],
+        messages: {
+            maximumExceeded: "Number of classes per file must not exceed {{ max }}"
+        }
+    },
+    create(context) {
+
+        const options = Object.assign(
+            { includeClassExpressions: true },
+            context.options[1] || {}
+        );
+        const maxClasses = context.options[0] || 1;
+
+        let classCount = 0;
+
+        return {
+            Program() {
+                classCount = 0;
+            },
+            "Program:exit"(node) {
+                if (classCount > maxClasses) {
+                    context.report({
+                        node,
```

Comment:
it make more sense to report the classes exceeding the limit (default to 1), rather than the program. thoughts?

Reply:
imo i would report the file still. we can never know which class is the "main class" and really its a per-file issue rather than per-class